### ISSUE TITLE
Sync projects and stations from nivabasen

### ISSUE THREAD

**Issue by jemmima** [NIVANorge/odm2-postgres-api#164]
This problem has been solved by @haakondr in observationsAPI as of 4853b276c6dd8a7b989c5cf75b0e70f371e0d358. To simplify the same solution will be used for odm2.

**Open questions:**
- [ ]  should we put a kafka_records table into odm2 in the way that's been done in observationsdb to keep track of project_stations that arrive before their corresponding station as so can't yet be inserted to samplingfeatures table? It will be simpler to maintain if we try to match the observations solution as much as possible - but maybe there is a cleaner solution for odm2 e.g. leaving the coordinates blank when creating site samplingfeature the first time and adding them when the matching station message appears.

**Comment by ZofiaRudjord** [comment]
There are pros and cons.
We could leave coordinate blank and allow to update coordinate only if they do not exist (yet).
So we do not risk "moving stations".

**Comment by jemmima** [comment]
> There are pros and cons.
> We could leave coordinate blank and allow to update coordinate only if they do not exist (yet).
> So we do not risk "moving stations".

I feel like the extra table is cleaner and will protect the integrity of the samplingfeatures table and avoid us having to have messy logic for updating that table. But this kafka_records table doesn't belong in the odm2 schema long term. I'm going to create the table for now while I figure out getting the kafka sink working and then we can discuss where it should live.

**Comment by jemmima** [comment]
At a table level the sync will look like this to start with:

[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBBW25pdmFiYXNlbi5wcm9qZWN0c10gLS0-IEIocHJvamVjdHMgdG9waWMpXG4gICAgQUFbbml2YWJhc2VuLm8tbnVtYmVyc10gLS0-IEIocHJvamVjdHMgdG9waWMpXG4gICAgQiAtLT4gRFtvZG0yLmRpcmVjdGl2ZXNdXG4gIFxuICAgIExbbml2YWJhc2VuLnByb2plY3Rfc3RhdGlvbnNdIC0tPk0ocHJvamVjdF9zdGF0aW9ucylcbiAgICBVW25pdmFiYXNlbi5zdGF0aW9uc10gLS0-VihzdGF0aW9ucyB0b3BpYylcbiAgICBNIC0tPiBFW29kbTIuc2FtcGxpbmdmZWF0dXJlc11cbiAgICBNIC0tPiBGW29kbTIuYWN0aW9uc11cbiAgICBWIC0tPiBFW29kbTIuc2FtcGxpbmdmZWF0dXJlc11cbiAgICBGIC0tPiBFXG4gICAgRiAtLT4gRCIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2UsImF1dG9TeW5jIjp0cnVlLCJ1cGRhdGVEaWFncmFtIjpmYWxzZX0)](https://mermaid-js.github.io/mermaid-live-editor/edit/##eyJjb2RlIjoiZ3JhcGggVERcbiAgICBBW25pdmFiYXNlbi5wcm9qZWN0c10gLS0-IEIocHJvamVjdHMgdG9waWMpXG4gICAgQUFbbml2YWJhc2VuLm8tbnVtYmVyc10gLS0-IEIocHJvamVjdHMgdG9waWMpXG4gICAgQiAtLT4gRFtvZG0yLmRpcmVjdGl2ZXM8XVxuICBcbiAgICBMW25pdmFiYXNlbi5wcm9qZWN0X3N0YXRpb25zXSAtLT5NKHByb2plY3Rfc3RhdGlvbnMpXG4gICAgVVtuaXZhYmFzZW4uc3RhdGlvbnNdIC0tPlYoc3RhdGlvbnMgdG9waWMpXG4gICAgTSAtLT4gRVtvZG0yLnNhbXBsaW5nZmVhdHVyZXNdXG4gICAgTSAtLT4gRltvZG0yLmFjdGlvbnNdXG4gICAgViAtLT4gRVtvZG0yLnNhbXBsaW5nZmVhdHVyZXNdXG4gICAgRiAtLT4gRVxuICAgIEYgLS0-IEQiLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)

Later we may add more attributes from nivabasen e.g. station attributes.

Proposed field mapping:

![Screenshot 2021-09-20 at 16 31 49](https://user-images.githubusercontent.com/18755759/134020346-709d03bb-6585-4071-9a16-afbf4ae3cce9.png)

Things I would like feedback on @ZofiaRudjord @haakondr @JamesSample :
- not currently planning to use the project start and end dates. I don't think odm2 needs to know about these, as we only will have projects for researchers who will have trouble finding their established stations without a project filter.
- linking projects with stations via the actions table - 'a project (directive) instructs that an action (project_station) is done at a site (samplingfeature)'. Would we lose any important information here?

**Comment by haakondr** [comment]
**station.geometryI**: think we should use the featuregeometry column, as this allows for GIS-based queries and indexing. I am sure @JamesSample has some thoughts on this. Storing original WKT does not hurt I though, but seems a bit redundant?

**project-station.id**: I am not sure what the usage of externalidentifiersystems.identifiersystemorganizationid, but seems to me like a column for storing stuff like "niva is id 123" and not identifying individual data rows?

**projects.onumbers**: may consider to use a postgres array type for onumbers column? I don't have a lot of experience with array types in postgres, or if there are any big drawbacks to using them. I guess the alternative is to comma-separate them or something? That works, but is a bit awkward if any of the contain the separator.

**fieldactivity**: If I understand this correctly, a project-station is modeled as a field activity action. That makes sense, but my initial thought was that an action maps to **one** field activity. Especially since the actions have time columns. We could perhaps make this a bit clearer with a different wording? not sure what that would be though, leaving that as an exercise to the reader -_-

**Comment by ZofiaRudjord** [comment]
> **project-station.id**: I am not sure what the usage of externalidentifiersystems.identifiersystemorganizationid, but seems to me like a column for storing stuff like "niva is id 123" and not identifying individual data rows?
> 
Look here: http://odm2.github.io/ODM2/schemas/ODM2_Current/diagrams/ODM2ExternalIdentifers.html
Each sampling features can have many external identifiers. I believe it fits our needs. The alternative is to store project-station.id as an annotation.

**Comment by ZofiaRudjord** [comment]
As far as I can see there is no mapping for metadata: createdBy for stations. I think it is an important attribute. Some researchers would like to explore there data by simply asking: give me all "my" data ( data connected to stations I created). I think this is the reason why this attribute was added.

**Comment by haakondr** [comment]
> **CreatedBy**: As far as I can see there is no mapping for metadata: createdBy for stations. I think it is an important attribute. Some researchers would like to explore there data by simply asking: give me all "my" data ( data connected to stations I created). I think this is the reason why this attribute was added.

My heart says **yes** but my brains say no :( currently there is no such information in nivabase, only in rare occations. All data written via aquamonitor is createdBy "WEB".

**Comment by ZofiaRudjord** [comment]
> > **CreatedBy**: As far as I can see there is no mapping for metadata: createdBy for stations. I think it is an important attribute. Some researchers would like to explore there data by simply asking: give me all "my" data ( data connected to stations I created). I think this is the reason why this attribute was added.
> 
> My heart says **yes** but my brains say no :( currently there is no such information in nivabase, only in rare occations. All data written via aquamonitor is createdBy "WEB".

Hm...Yes, in aquamonitor is it WEB (it does not matter you so much because you only use it to visualize and export, not to explore) but in nivabasen it is filled. At least Jens suggested to me to explore data this way.

**Comment by haakondr** [comment]
reading this 3 or 4 times, I think it could work. Still a bit confusing, but that we can solve with documentation.

**Comment by ZofiaRudjord** [comment]
> **fieldactivity**: If I understand this correctly, a project-station is modeled as a field activity action. That makes sense, but my initial thought was that an action maps to **one** field activity. Especially since the actions have time columns. We could perhaps make this a bit clearer with a different wording? not sure what that would be though, leaving that as an exercise to the reader -_-
>
I am not sure I understand you concern here ( or is there any concern?). Can you please elaborate?
I understand the logic like this: Project owner takes an action of going to the field and establishing/connect a station. Some time later it goes to the field again (different location) and connects(if already exists)/establish another stations...
Would that not work?

**Comment by ZofiaRudjord** [comment]
**Project start and end date**: I am not sure how important it is for data exploration...Perhaps @JamesSample  knows?

**Comment by haakondr** [comment]
> > **fieldactivity**: If I understand this correctly, a project-station is modeled as a field activity action. That makes sense, but my initial thought was that an action maps to **one** field activity. Especially since the actions have time columns. We could perhaps make this a bit clearer with a different wording? not sure what that would be though, leaving that as an exercise to the reader -_-
> 
> I am not sure I understand you concern here ( or is there any concern?). Can you please elaborate?
> I understand the logic like this: Project owner takes an action of going to the field and establishing/connect a station. Some time later it goes to the field again (different location) and connects(if already exists)/establish another stations...
> Would that not work?

It would work. I was mostly confused by the wording "fieldactivity". Not sure what would be best here, but something like "registerStation" or something like that?

**Comment by ZofiaRudjord** [comment]
I think it would be valuable to bring @jeved onboards this nivabase->odm2 project-station mapping, to make sure we do not overlook important attributes.

**Comment by haakondr** [comment]
> I think it would be valuable to bring @jeved onboards this nivabase->odm2 project-station mapping, to make sure we do not overlook important attributes.

There is always a possibility of rerunning consumer to fill in more attributes at a later stage.

**Comment by ZofiaRudjord** [comment]
> > > **fieldactivity**: If I understand this correctly, a project-station is modeled as a field activity action. That makes sense, but my initial thought was that an action maps to **one** field activity. Especially since the actions have time columns. We could perhaps make this a bit clearer with a different wording? not sure what that would be though, leaving that as an exercise to the reader -_-
> > 
> > 
> > I am not sure I understand you concern here ( or is there any concern?). Can you please elaborate?
> > I understand the logic like this: Project owner takes an action of going to the field and establishing/connect a station. Some time later it goes to the field again (different location) and connects(if already exists)/establish another stations...
> > Would that not work?
> 
> It would work. I was mostly confused by the wording "fieldactivity". Not sure what would be best here, but something like "registerStation" or something like that?
>
Yes, odm2 controlled vocabularies are often  bit abstract. We could update actiontype_cv. Would you like to go down this road? ( update cv's)

**Comment by haakondr** [comment]
I think it could make sense that we define our own controlled vocabularies for niva-specific concepts. I have made a script which downloads CV's to json (https://github.com/NIVANorge/odm2-postgres-api/blob/dev/src/odm2_postgres_api/controlled_vocabularies/download_cvs.py). These json files are read on API startup and loaded into odm2 database.

Would be easy to append this json with new CV's. However, I think the download_cvs.py script will overwrite our own CV's in its current state, so that should be fixed.

**Comment by ZofiaRudjord** [comment]
> > I think it would be valuable to bring @jeved onboards this nivabase->odm2 project-station mapping, to make sure we do not overlook important attributes.
> 
> There is always a possibility of rerunning consumer to fill in more attributes at a later stage.

True, but I think it is good to come up with the best we can as of today. Of course if we discover later that we missed out sth we can re-run. But it will always be associated with extra effort.

**Comment by jemmima** [comment]
I had planned to do our best with a proposal before running it past @jeved

**Comment by jemmima** [comment]
> **project-station.id**: I am not sure what the usage of externalidentifiersystems.identifiersystemorganizationid, but seems to me like a column for storing stuff like "niva is id 123" and not identifying individual data rows?

@haakondr  you're right that `externalidentifiersystems.identifiersystemorganizationid` represents 'niva' for example, but `samplingfeatureexternalidentifiers.samplingfeatureexternalidentifier` is for the individual id - I had the arrow going to the wrong place in the diagram, fixed now.

**Comment by JamesSample** [comment]
> station.geometryI: think we should use the featuregeometry column, as this allows for GIS-based queries and indexing. I am sure @JamesSample has some thoughts on this. Storing original WKT does not hurt I though, but seems a bit redundant?

I agree about filling in the `featureGeometry` column so that we can use the PostGIS spatial functions. Also agree that having both `geometry` and WKT is a bit redundant, although not a big deal as long as they can be kept consistent. My default preference would be to store as `geometry` in the database and then convert to WKT, WKB etc. as required. Note that PostGIS optionally implements "extended" WKT ([EWKT](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Format_variations)), where the SRS is embedded within the text string. I find this really useful for passing spatial data around, so if we store anything in the `FeatureGeometryWKT` column I'd recommend EWKT (assuming this doesn't cause problems elsewhere - not all libraries support it).

> Project start and end date: I am not sure how important it is for data exploration...Perhaps @JamesSample knows?

I don't think querying by project start and end date is very important (although I'm sure some users would if given the option). In general, the start and end date from a data perspective can be queried easily as just the period during which sampling took place. The project start and end dates are presumably stored in Instipro, and I don't think it's necessary to have them in our database too.

**Comment by jemmima** [comment]
> > station.geometryI: think we should use the featuregeometry column, as this allows for GIS-based queries and indexing. I am sure @JamesSample has some thoughts on this. Storing original WKT does not hurt I though, but seems a bit redundant?
> 
> I agree about filling in the `featureGeometry` column so that we can use the PostGIS spatial functions. Also agree that having both `geometry` and WKT is a bit redundant, although not a big deal as long as they can be kept consistent. My default preference would be to store as `geometry` in the database and then convert to WKT, WKB etc. as required. Note that PostGIS optionally implements "extended" WKT ([EWKT](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Format_variations)), where the SRS is embedded within the text string. I find this really useful for passing spatial data around, so if we store anything in the `FeatureGeometryWKT` column I'd recommend EWKT (assuming this doesn't cause problems elsewhere - not all libraries support it).

@JamesSample should the value in "geometry" in the station topic be inserted into the `featuregeometry` column as is? i.e. `POINT(10.6,90.67)`

**Comment by JamesSample** [comment]
@jemmima 

When you insert, you convert from (E)WKT to PostGIS geometry using one of the PostGIS spatial functions e.g.:

`INSERT INTO myspatial_table(geom) VALUES(ST_GeomFromText('POINT(11, 60)', 4326));`

or

`INSERT INTO myspatial_table(geom) VALUES(ST_GeomFromEWKT('SRID=4326; POINT(11, 60)'));`

(see [here ](https://postgis.net/docs/ST_GeomFromText.html) and [here ](https://postgis.net/docs/ST_GeomFromEWKT.html) for details). You can then convert back to (E)WKT when required using `ST_AsText` ([here](https://postgis.net/docs/ST_AsText.html)) or `ST_AsEWKT` ([here](https://postgis.net/docs/ST_AsEWKT.html)).

In all these functions, it's easiest to use [EPSG codes](https://epsg.io/) to specify the SRIDs (as in the examples above).

**Comment by jemmima** [comment]
@JamesSample - thanks! I wasn't familiar with the PostGIS geometry. Will all the coordinates in nivabasen be EPSG=4326? Or is that specified in a field we need to get from nivabasen that isn't currently in the station topic?

**Comment by JamesSample** [comment]
@jemmima 

Hopefully @roarbra can confirm, but from a quick look _almost_ all the co-ordinates in `niva_geometry.sample_points` are entered as latitude and longitude. It's not explicitly stated, but these values are most likely based on the WGS84 ellipsoid, which is EPSG 4326. The other obvious possibility would be ETRS89, which is EPSG 4258. @roarbra or @jankarud might be able to confirm which has been used, but for practical purposes the difference between them is negligible, so I'd be happy just assuming everything is 4326.

There are 14 records in `niva_geometry.sample_points`  where latitude and longitude are `NULL`, but looking at the columns `EPSG_SYS`, `x` and `y`, the numbers don't make much sense, so I'd ignore these for now.

Are you ingesting data from elsewhere too, or is it mainly from `niva_geometry.sample_points`? There are a few other common co-ordinate systems that you might come across at NIVA (mostly UTM-based), but it doesn't look like you need to worry about them if all the data comes from `niva_geometry.sample_points`.

**Comment by jemmima** [comment]
> @jemmima
> 
> Hopefully @roarbra can confirm, but from a quick look _almost_ all the co-ordinates in `niva_geometry.sample_points` are entered as latitude and longitude. It's not explicitly stated, but these values are most likely based on the WGS84 ellipsoid, which is EPSG 4326. The other obvious possibility would be ETRS89, which is EPSG 4258. @roarbra or @jankarud might be able to confirm which has been used, but for practical purposes the difference between them is negligible, so I'd be happy just assuming everything is 4326.
> 
> There are 14 records in `niva_geometry.sample_points` where latitude and longitude are `NULL`, but looking at the columns `EPSG_SYS`, `x` and `y`, the numbers don't make much sense, so I'd ignore these for now.
> 
> Are you ingesting data from elsewhere too, or is it mainly from `niva_geometry.sample_points`? There are a few other common co-ordinate systems that you might come across at NIVA (mostly UTM-based), but it doesn't look like you need to worry about them if all the data comes from `niva_geometry.sample_points`.

Thanks James! If this is modelled in nivabasen we should probably not make this assumption, as new data could be added to nivabasen during the lifetime of this sync which specifies a different EPSG, it would be hard for us to track that in future. @haakondr or @nmarquesin can confirm where the data on kafka is coming from - I only have access to the kafka topics at the moment and would have to dig into the that codebase to find out where it's coming from in nivabasen. If we need EPSG to ingest to odm2 we need to modify the kafka topics to include that from nivabasen I think.

**Comment by roarbra** [comment]
The fields longitude / latitude should be **EPSG:4326.**

When entering coordinates we will use the fields: EPSG_SYS, X, Y. A trigger on the table will call a web service that sets the longitude, latitude-fields, and in addition fetches results for STATION_ATTRIBUTES.

**Comment by JamesSample** [comment]
Thanks Roar!

@jemmima That probably explains the 14 NULLs: cases where users have entered invalid co-ordinates and the web service has failed to convert them to EPSG 4326.

**Comment by haakondr** [comment]
currently all the data on kafka are coming from niva_geometry.sample_points. 

I know that some of the begroing begroing stations in access do store locations in various coordinate systems. Could be a good idea to store the "source" EWKT string in cases where we want to map back to a different coordinate system? but then again, this information should be available in the original kafka message, so I don't see any reason to put this in the odm2 table as well.

**Comment by JamesSample** [comment]
You will need to specify the SRID when you insert data into the database and, in general, it's a bad idea to mix SRSs within the same geometry column. If you're getting data from different sources/SRSs, I'd recommend converting everything to EPSG 4326 and inserting like that. From what Roar says above, values in `niva_geometry.sample_points` are already in EPSG 4326, so no conversion is necessary. 

Co-ordinate transformations can be "lossy", but the errors are usually negligible for NIVA applications. As long as the original (E)WKT is available somewhere in the kafka message, I agree with Håkon that we don't need to store it in ODM2.

**Comment by jemmima** [comment]
> should we allow upserts on directives table? Don't think we need to make a whole new project every time a new onumber is added in nivabasen on the same project, which seems to happen annually for multi-year projects)

@haakondr  @ZofiaRudjord  I propose we make it so only the onumbers column can be updated on conflict with directivename? Depends on whether project name is unique in nivabasen - I don't have access to these tables. Will ask Jens but if either of you know off hand that would be helpful to know when making this decision.

**Comment by jeved** [comment]
@jemmima you are right, projects_name_un is an unique constraint on projects.project_name. The naming of the constraints in nivabase can be poor, so better check the definition:

... TABLE NIVADATABASE.PROJECTS ADD CONSTRAINT PROJECTS_NAME_UN UNIQUE (PROJECT_NAME )...

![bilde](https://user-images.githubusercontent.com/27200772/135216206-ce4acd63-58bf-4812-8d7a-736832b79da9.png)

**Comment by ZofiaRudjord** [comment]
In nivabase project table does not have O numbers, while we attempt to store Onumbers in the same table as general project info. Hence I would go for your original suggestion @jemmima to update Onumbers on conflict ( that is to allow to append to the list).

**Comment by haakondr** [comment]
## Project and station nivabase id references

do we have a link between the following:
  - nivabase.station and odm2.samplingfeature ?
  - nivabase.projects and odm2.directives ?

project stations are linked via samplingfeatureexternalidentifiers.samplingfeatureexternalidentifier but we should perhaps do the same for stations/projects? A lot of nivabase/aquamonitor functionality relies on linking stuff via their project_id/station_id so I think it would be useful to have this stored in odm2 to enable good integrations.

Another perspective is how to deal with updates. Example:

project name is updated in nivabase, we want to update project name in odm2 (if that is desired behaviour). In order to do so we need a solid identifier in odm2. The externalidentifiersystems does kinda work, but does not look like directives are setup with this. Alternatively, we could perhaps utilise annotations for this?

## Multiple project stations for one station

Did we conclude on solving multiple station codes for one station? Sorry if I've missed this, but I need a refresher :)

Some alternatives:
- One samplingfeature per project station and drop unique constraint on station code?

the drawback of this approach is that we get a lot of duplicate samplingfeatures with the same location, would probably need some filtering to join them together in a map visualization

- I see that samplingfeatureexternalidentifiers has a one:many relationship with samplingfeatures so station codes / names could potentially be modelled using this table. Has this been explored?

Would need to either drop the usage of samplingfeaturecode or elect a "primary code" for it?  

.. or did we discuss other alternatives?

**Comment by roarbra** [comment]
> 
> * I see that samplingfeatureexternalidentifiers has a one:many relationship with samplingfeatures so station codes / names could potentially be modelled using this table. Has this been explored?
> 

This seems like exactly what you need to represent the behavior of Nivadatabase.

**Comment by ZofiaRudjord** [comment]
> project name is updated in nivabase, we want to update project name in odm2 (if that is desired behaviour). In order to do so we need a solid identifier in odm2. The externalidentifiersystems does kinda work, but does not look like directives are setup with this. Alternatively, we could perhaps utilise annotations for this?

I am not convinced that directives ( linked via actions to sampling features) is a good way of storing niva_projects. Perhaps we could add a table: niva_directives and the relevant links to extrenalidentifiers? 

Several of us ( including @JamesSample  and @jemmima  ) have discussed that odm2 is a good fit but where it is not, we should modify it to serve niva data needs. We could agree on the convention that e.g. when we add a table we prepend name niva_ before it.

**Comment by haakondr** [comment]
I started on a branch off feature/kafka-consumer and setup a unit test environment which should serve as a good basis for doing changes: https://github.com/NIVANorge/odm2-postgres-api/tree/haakon-kafka (currently in conflict with dev, same as feature/kafka-consumer but I'll sort that out).

I suggest we sit down next week when @jemmima  is back (?) and discuss the data model details and continue on the implementations (see my [questions above](https://github.com/NIVANorge/odm2-postgres-api/issues/164#issuecomment-984494248)

**Comment by haakondr** [comment]
Summarized discussion with @jemmima, @ZofiaRudjord  and myself today.

We propose to model projects, project stations and stations as the following:

- nivabase.stations -> odm2.samplingfeatures
  - propose to leave samplingfeaturecode as nullable for now *
- nivabase.projects -> odm2.projects (new table)
- nivabase.project stations -> odm2.project_stations (new table)
- references to nivabase id's are done with `externalidentifiersystems`
  - do we make one "externalidentifiersystem" for nivabase, or one for each table? we could do externalidentifiersystemname="nivabase.projects_stations" for instance. 
- All modifications to odm2 schema should be documented in a separate document. We do have https://github.com/NIVANorge/odm2-postgres-api/blob/dev/src/odm2_postgres_api/db_init/odm2_changelog.md already, but this should probably be written as a "data model" document 

* we discussed various possibilities for filling in samplingfeaturecode. Problem is that a station in nivabase may have multiple project stations and station codes. Also, project station codes only unique within a project. Therefore, the project station codes cannot be used as samplingfeaturecode without changes. 

On the topic of progressing towards a unique station code, we agreed to scope it out for now, but it would be beneficial to have it.

**Comment by ZofiaRudjord** [comment]
If we have one "externalidentifiersystem" for nivabase and use externalidentifiersystemname="nivabase.projects_stations", 
which column would you store station_id ? Would you extend the table?

**Comment by jemmima** [comment]
the station id would be in samplingfeatureexternalidentifiers, no?
<img width="989" alt="Screenshot 2021-12-08 at 10 31 04" src="https://user-images.githubusercontent.com/18755759/145184270-5d4b551d-6d5f-45a9-9e26-3b16796a3f3a.png">

**Comment by jemmima** [comment]
@ZofiaRudjord we need to consider all the changes we need to make to existing systems if we go with this approach - currently samplingfeaturecode is used for a lot.

**Comment by ZofiaRudjord** [comment]
Sorry I missed that link somehow ( probably not enough coffee yet :D) 

samplingfeaturecode is used a lot for sites for loggers. It is used extensively for samples too.
Perhaps for nivabase site sampleingfetaurecode could be set to the same as nivabase_samplingfeatureid?

**Comment by ZofiaRudjord** [comment]
Another point is that it should be possible to get the site by station_project code and project.

**Comment by haakondr** [comment]
but could we not use samplingfeatureuuid in our logger ingestion?

I don't think we need our ingestion pipeline to be all generic, after all it is not that many data sources. We could even just write a mapping station code -> samplingfeatureuuid hardcoded in the ingestion pipeline. 

For new data sources we could configure the logger software to send the uuid instead of station code.

**Comment by ZofiaRudjord** [comment]
But then you need a mapping being available in grafana too. 
It is much easier to work with human readable identifiers. You also complaint about it for ferrybox data.

**Comment by jemmima** [comment]
We already have to change all the systems using odm2 as they have the station name currently coming from samplingfeaturecode - so we can't get around doing this work if we take this approach to the syncing

**Comment by ZofiaRudjord** [comment]
Mass spec operates on samples. But you worry is that it is not robust enough to rely on uniqueness of samplingfeaturecode for specimen since it is not enforced? I agree to that. Thats why I suggested to put nivabase_samplingfeatureid for sites coming from nivabase to keep the unique samplingfeaturecode constraint (and not make it nullable)

**Comment by ZofiaRudjord** [comment]
To avoid going in circles I would like to summarise two main syncing strategies we have discussed.
**Option 1** Use samplingfeatures table as a joint of stations and project_stations from nivabasen. Create project table with one - to -one mapping with nivabase station. Samplingfeature codes are unique identifiers and are created as concatenate of project_station codes and project. 
Advantages:
* This strategy is align with current odm2 querry functions which rely on samplingfeaturecodes as unique human readable (and meaningful) identifiers. 
* It would be little work to update existing code for loggers/mass_sepc/spectral(cdom) endpoints and functionality in the corresponding code bases.
Drawback/issues:
* Kafka sink part would be different then the one in observation-db -> less code re-usability, more coding needed.
**Option 2** Use samplingfeatures table as the Station table in the nivabasen. Create project_station and project tables.
Advantages:
* Could re-use kafka-sink from observation-db
Drawback/issues:
* Creation of unique samplingfeaturecode for stations would be problematic 
* It would be more work to update loggers/mass_sepc/spectral(cdom) endpoints which would have to querry 3 tables : project, proejct_stations and samplingfeatures to get station for relevant combination of (project_staion code and project name).

**Comment by ZofiaRudjord** [comment]
My personal vote goes to Option 1. But it is your lead @haakondr .

**Comment by haakondr** [comment]
Thanks for the summary. I agree that those two are the best options. I have a question about **option 1**:

- how do we deal with new project stations?

Example: we have station A with 3 project stations (A, B and C). At the time of sync we concatenate their station codes into one A,B,C. Then a user creates a fourth project station D attached to the station. Should we then update the samplingfeaturecode to A,B,C,D ?

If I've understood the use case correctly, we want to refer to stations by a unique station code. If we have code/systems referring to code A,B,C then things would not work.

**Comment by ZofiaRudjord** [comment]
Perhaps It was not clear what I meant by a joint of project station and station. My idea was (  the same as the initial idea described by @jemmima ) that if we have 3 project stations we will create 3 samplingfeatures. That is A, B, and C in your example. If you get a new one you register it a a new samplingfeature D.

**Comment by jemmima** [comment]
@haakondr I think the consumer is functioning well now - we aren't yet handling delete messages but we don't need that to make it available to Zofia and Liza, so I can add that after. The only open question I have before we go to prod is whether to remove the uniqueness constraint from projects.projectname, as there were a couple of violations. I'm trying to access nivabasen now to check if the conflicts are valid or not.
